### PR TITLE
Restriction SemiCompleted

### PR DESCRIPTION
**Pull request by Sothe5**

### CODE REVIEW

File: prLogicalElements/Operation.java [hunk 1]
```diff
@@ -64,13 +60,37 @@ public boolean isLiteral() {
...
 		for(Element element: this.components) {
-			if (!element.isLiteral()) {
+			if (element instanceof Operation) {
 				operations.add((Operation) element);
 			}
 		}
 		return operations;
 	}
 	
+	public Set<Literal> getInmediateLiterals() {
```
Comment (zolastro) [inline prLogicalElements/Operation.java]: Maybe we should extract this function and `getLiterals()` into a new class with static operations? What do you think?

File: prLogicalElements/Operation.java [hunk 2]
```diff
@@ -64,13 +60,37 @@ public boolean isLiteral() {
...
+		for (Element elements : this.components) {
+			if (elements instanceof Literal) {
+				literals.add((Literal) elements);
+			}
+		}
+		
+		return literals;
+	}
+	
+	public List<Operation> getInmediateOperations() {
```
Comment (zolastro) [inline prLogicalElements/Operation.java]: This is exactly the same as `getOperations()`, isn't it?

File: Tas/Restrict.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			restriction(operation);
+			restrictForward(operation.getInmediateOperations());
+		}
+	}
+
+	private static void restriction(Operation operation) {
+		Set<Literal> literalsOfRoot = operation.getInmediateLiterals();
+		List<Operation> operationsOfRoot = operation.getInmediateOperations();
+
+		restrictWithRoot(literalsOfRoot, operationsOfRoot);
```
Comment (zolastro) [inline Tas/Restrict.java]: You could just pass the root operation and extract in `restrictWithRoot` the rest of the arguments.

File: Tas/Restrict.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		restrictComponents(operationsOfRoot);
+	}
+
+	private static void restrictWithRoot(Set<Literal> literalsOfRoot, List<Operation> operationsOfRoot) {
+		for (Literal literals : literalsOfRoot) {
+			boolean canRestrict = true;
+			for (Operation operations : operationsOfRoot) {
+				if (!operations.getInmediateLiterals().contains(literals)) {
+					canRestrict = false;
+					// probar si el break rompe los dos bucles o solo este (quiero el solo este)
```
Comment (zolastro) [inline Tas/Restrict.java]: It only breaks the inner loop. Check [this](https://stackoverflow.com/questions/886955/breaking-out-of-nested-loops-in-java) out.

File: prLogicalElements/Operation.java (hunk 2 above)
Comment (Sothe5) [reply prLogicalElements/Operation.java]: Yes, it does the same but why did you do a method called `getLiterals()` that get All the literals in the tree and then a method called `getOperations()` that only get the operations of the first level XD.
ps: i didn't even read the second method just assume that does the same as `getLiterals()` but with operations

File: Tas/Restrict.java (hunk 3 above)
Comment (Sothe5) [reply Tas/Restrict.java]: Yes I could do that and the same in restrictComponents() but then i will have to repeat two times the extraction of the components and as that method is private i don't see the problem of having one argument more in order to not have redundancy.

File: Tas/Restrict.java (hunk 4 above)
Comment (Sothe5) [reply Tas/Restrict.java]: ok then i will add that break; that reduce the innecesary proccesses.

File: prLogicalElements/Operation.java (hunk 1 above)
Comment (Sothe5) [reply prLogicalElements/Operation.java]: Maybe but i dont see it neccesary yet if we add some more methods that works with preprocessing information for other jobs could be a nice idea.

**Pull request merged**